### PR TITLE
New completions

### PR DESCRIPTION
**Pull request by jansorg**
- Use new completions endpoint
- Add support for nested completions, i.e. display nested completions with indentation
- Use Sublime's placeholders to implement Kite's

To do:
- [x] send a completion request when a placeholder is selected/activated
- [x] Remove debug logger output

Limitations:
- with `auto_complete_commit_on_tab=false` (i.e. the default) pressing `TAB` when completions are shown for a placeholder  will not insert the currently selected option. Sublime seems to be calling `insert_best_completion` only if there's no placeholder in the document. This means that it's working in the way as it's supposed to be but it's still a limitation. Use `auto_complete_commit_on_tab=true` to work around this. 

Known issues:
- [ ] Sublime is filtering the returned completions with the current prefix. I couldn't find a way to force Sublime to show completion items nevertheless. This means that we have to include the prefix in the completions returned by Kite.
- [ ] Sometimes the completions are cut off, i.e. they aren't displayed completely. This is a Sublime issue, as far as I can tell: https://github.com/SublimeTextIssues/Core/issues/633
- [ ] replace the range given in the response when a completion is inserted.
So far I wasn't able to find a way to implement this. The range is attached to completion items and Sublime doesn't seem to offer a way to be notified which particular item was selected by the user. https://forum.sublimetext.com/t/call-function-when-completion-selected-on-query-completions/7492/2 , https://github.com/SublimeTextIssues/Core/issues/2754. I don't see a reliable way to work around this. 
Theoretically it's possible to guess the inserted item when `commit_completion` is called. But this is unreliable because 
        a) the user may have modified the original item's content by updating placeholders 
        b) there could be nested placeholders, i.e. placeholders inserted for an inserted placeholder 
        c) we'd have to guess what Sublime has already removed when it inserted the completion
We may have *some* success here but it won't be reliable enough, I think.

### CODE REVIEW

File: lib/handlers.py [hunk 1]
```diff
@@ -278,18 +291,55 @@ def _run_auto_complete(view):
...
+    def _flatten_completions(self, completions, nesting=0):
+        if not completions:
+            return []
+
+        result = []
+        for c in completions:
+            result.append((self._brand_completion('  ' * nesting + c['display'], c['hint']),
+                           self._placeholder_text(c)))
+            if 'children' in c:
+                result += self._flatten_completions(c['children'], nesting + 1)
```
Comment (its-dhung) [inline lib/handlers.py]: I believe using `result.extend(self._flatten_completions(c['children'], nesting + 1))` should be more performant here.

File: lib/handlers.py [hunk 2]
```diff
@@ -278,18 +291,55 @@ def _run_auto_complete(view):
...
+
+    @staticmethod
+    def _placeholder_text(completion):
+        text = completion['snippet']['text']
+        try:
+            placeholders = completion['snippet']['placeholders']
+            # sort placeholders in reverse order for easier string patching
+            # we assume that placeholders do not overlap
+            copy = list(placeholders)
+            copy.sort(key=lambda i: i['begin'], reverse=True)
```
Comment (its-dhung) [inline lib/handlers.py]: A more compact to write this is:
```python
copy = sorted(placeholders, key=lambda i: i['begin'], reverse=True)
```

File: lib/handlers.py [hunk 3]
```diff
@@ -230,16 +227,32 @@ def on_query_completions(self, view, prefix, locations):
...
-                     c['insert']) for c in cls._received_completions
-                ]
+                completions = self._flatten_completions(cls._received_completions)
 
             cls._received_completions = []
             cls._last_location = None
 
             return completions
 
+    def on_post_text_command(self, view, command_name, args):
```
[the remark this answers is not in the document]
Comment (jansorg) [reply lib/handlers.py]: Which `_handle` method are you referring to? There no such method in `CompletionsHandler`.

File: lib/handlers.py [hunk 4]
```diff
@@ -230,16 +227,32 @@ def on_query_completions(self, view, prefix, locations):
...
+        if command_name not in ('prev_field', 'next_field', 'commit_completion', 'insert_best_completion'):
+            logger.debug("skipping post text command {}".format(command_name))
```
[the remark this answers is not in the document]
Comment (jansorg) [reply lib/handlers.py]: I've removed both calls to logging

File: lib/handlers.py [hunk 5]
```diff
@@ -230,16 +227,32 @@ def on_query_completions(self, view, prefix, locations):
...
+            return
+        if len(view.sel()) != 1:
+            logger.debug("skipping post text command {} with != 1 selections".format(command_name))
```
[the remark this answers is not in the document]
Comment (jansorg) [reply lib/handlers.py]: Removed

File: lib/handlers.py (hunk 1 above)
Comment (jansorg) [reply lib/handlers.py]: Good point, I'll change it

File: lib/handlers.py (hunk 2 above)
Comment (jansorg) [reply lib/handlers.py]: Thanks

File: lib/handlers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (its-dhung) [reply lib/handlers.py]: My mistake, I misread this file. The placement here is fine!

File: lib/handlers.py [hunk 6]
```diff
@@ -278,18 +289,54 @@ def _run_auto_complete(view):
             'next_completion_if_showing': False,
         })
 
+    def _flatten_completions(self, completions, nesting=0):
+        if not completions:
+            return []
+
+        result = []
+        for c in completions:
+            result.append((self._brand_completion('  ' * nesting + c['display'], c['hint']),
```
Comment (its-dhung) [inline lib/handlers.py]: We should remove the indentation from the nesting for Sublime.

Comment (jansorg) [reply lib/handlers.py]: I've removed the indentation and added the new settings.
@its-dhung

**Pull request merged**